### ISSUE TITLE
Setting Response headers in Java is not threadsafe

### ISSUE THREAD

**Issue by bbarkley** [playframework/playframework#8490]
### Play Version
All

### API
Java

### Expected Behavior

Headers set on the `Http.Response` should be output with the response, regardless of where they're being set.

### Actual Behavior

The Map used for headers in `Http.Response` is a `TreeMap` which is not threadsafe. If headers are set concurrently by multiple threads it's possible that the map will be corrupt, and not all data will be retained. We have observed this when a header is set in an Action outside of the resolution of the CompletionStage. The code was roughly (we're still on 2.4):

```
public abstract class HeaderSettingAction<T> extends Action<T> {
  @Override
  public final F.Promise<Result> call(Http.Context ctx) throws Throwable {
    F.Promise<Result> result = delegate.call(ctx);
    ctx.response().setHeader(“foo", "1");
    return result;
  }
}
```
A header was also being set in a Controller, and about 1/200000 requests the header set in the controller was not present in the response. In this case it's easy enough to move the set header into a `.map` to guarantee ordering, but since the current Java API uses shared ThreadLocal state this would be easy enough to do accidentally in multiple CompletionStages running in parallel. Since this is a very subtle and hard to diagnose bug I'd suggest fixing it in Play.

I would expect that the map for the headers in`Http.Response` could be replaced with a `ConcurrentSkipListMap` to resolve this issue. I can put together a PR if desired.

### Reproducible Test Case

Since this is a negative test case of a concurrency issue I haven't supplied a test case. If desired I can work on one, but I would expect that it would add significant running time to the test suite since this is a rare bug.

**Comment by mkurz** [comment]
This will fix itself starting with Play 2.7 where `Http.Context` is deprecated.

**Comment by mkurz** [comment]
Closing. The http context ThreadLocal was removed in Play 2.8. Play <2.7 is not supported anymore.
